File: sdk/typescript/src/workflows-static.ts
import { writeFileSync } from "node:fs";
import YAML from "yaml";

import type {
  WorkflowDefinitionSpec,
  WorkflowStep,
  WorkflowStepAppCall,
} from "./providers/workflow.ts";

type StaticWorkflowDefinitions = {
  definitions: Array<{
    id?: string;
    steps: Array<{
      app: string;
      operation?: string;
    }>;
  }>;
};

/**
 * Serializes workflow app-call steps for provider packaging.
 */
export function workflowsToYaml(
  specs: readonly WorkflowDefinitionSpec[],
): string {
  const definitions = specs
    .map((spec) => ({
      id: spec.id,
      steps: workflowAppCallsFromTarget(spec.target),
    }))
    .filter((definition) => definition.steps.length > 0);
  const payload: StaticWorkflowDefinitions = { definitions };
  return YAML.stringify(payload);
}

/**
 * Writes workflow app-call metadata to disk as YAML.
 */
export function writeWorkflowsYaml(
  path: string,
  specs: readonly WorkflowDefinitionSpec[],
): void {
  const yaml = workflowsToYaml(specs);
  if (!yaml.trim()) {
    return;
  }
  writeFileSync(path, yaml, "utf8");
}

function workflowAppCallsFromTarget(
  target: WorkflowDefinitionSpec["target"],
): Array<{ app: string; operation?: string }> {
  if (!target?.steps?.length) {
    return [];
  }
  const out: Array<{ app: string; operation?: string }> = [];
  for (const step of target.steps) {
    const appCall = workflowStepAppCall(step);
    if (!appCall?.name?.trim()) {
      continue;
    }
    out.push({
      app: appCall.name.trim(),
      operation: appCall.operation?.trim() || undefined,
    });
  }
  return out;
}

function workflowStepAppCall(step: WorkflowStep): WorkflowStepAppCall | undefined {
  if (step.app?.name) {
    return step.app;
  }
  if (step.action?.case === "app") {
    return step.action.value;
  }
  return undefined;
}

File: sdk/go/workflows_static.go
package gestalt

import (
	"fmt"
	"os"
	"strings"

	"gopkg.in/yaml.v3"
)

const envWriteWorkflows = "GESTALT_APP_WRITE_WORKFLOWS"

type staticWorkflowDefinitions struct {
	Definitions []staticWorkflowDefinition `yaml:"definitions,omitempty"`
}

type staticWorkflowDefinition struct {
	ID    string                 `yaml:"id,omitempty"`
	Steps []staticWorkflowAppCall `yaml:"steps,omitempty"`
}

type staticWorkflowAppCall struct {
	App       string `yaml:"app"`
	Operation string `yaml:"operation,omitempty"`
}

func writeDeclaredWorkflowsYAML(provider any, path string) error {
	decl, ok := provider.(interface {
		DeclaredWorkflowDefinitions() ([]WorkflowDefinitionSpec, error)
	})
	if !ok {
		return nil
	}
	specs, err := decl.DeclaredWorkflowDefinitions()
	if err != nil {
		return fmt.Errorf("declare workflow definitions: %w", err)
	}
	workflows := staticWorkflowDefinitionsFromSpecs(specs)
	if len(workflows.Definitions) == 0 {
		return nil
	}
	if err := ensureOutputDir("workflows", path); err != nil {
		return err
	}
	data, err := yaml.Marshal(workflows)
	if err != nil {
		return fmt.Errorf("encode workflows yaml: %w", err)
	}
	if err := os.WriteFile(path, data, 0o644); err != nil {
		return fmt.Errorf("write workflows yaml %q: %w", path, err)
	}
	return nil
}

func staticWorkflowDefinitionsFromSpecs(specs []WorkflowDefinitionSpec) staticWorkflowDefinitions {
	if len(specs) == 0 {
		return staticWorkflowDefinitions{}
	}
	out := staticWorkflowDefinitions{Definitions: make([]staticWorkflowDefinition, 0, len(specs))}
	for _, spec := range specs {
		steps := staticWorkflowAppCallsFromTarget(spec.Target)
		if len(steps) == 0 {
			continue
		}
		out.Definitions = append(out.Definitions, staticWorkflowDefinition{
			ID:    strings.TrimSpace(spec.ID),
			Steps: steps,
		})
	}
	return out
}

func staticWorkflowAppCallsFromTarget(target *BoundWorkflowTarget) []staticWorkflowAppCall {
	if target == nil || len(target.Steps) == 0 {
		return nil
	}
	out := make([]staticWorkflowAppCall, 0, len(target.Steps))
	for _, step := range target.Steps {
		if step.App == nil {
			continue
		}
		appName := strings.TrimSpace(step.App.Name)
		if appName == "" {
			continue
		}
		out = append(out, staticWorkflowAppCall{
			App:       appName,
			Operation: strings.TrimSpace(step.App.Operation),
		})
	}
	return out
}
